tests/test_smart_ptr.cpp

```diff
@@ -367,4 +361,32 @@ TEST_SUBMODULE(smart_ptr, m) {
                 list.append(py::cast(e));
             return list;
         });
+
+    // test_holder_mismatch
+    // Tests the detection of trying to use mismatched holder types around the same instance type
+    struct HeldByShared {};
+    struct HeldByUnique {};
+    // HeldByShared declared with shared_ptr holder, but used with unique_ptr later
+    py::class_<HeldByShared, std::shared_ptr<HeldByShared>>(m, "HeldByShared");
+    m.def("register_mismatch_return", [](py::module m) {
```

Comment:
My brief reading of this PR leads me to believe that there's a sharp edge:  a user could still hit a segfault if they call `py::cast<mismatched_holder_type>(my_object)`. Totes get the idea of wanting to shift overhead from `py::cast<>()` (runtime) to function declarations (binding-time, e.g. `.def()`), but it now means there are now distinct entry points to this type-checking that users can trigger :(

Possible resolutions:

- if it's worth the risk, then just mention it in the holder documentation (as a diff in this PR)
- If it's not worth the risk, somehow enable raw `py::cast<>()` to do a runtime check. This may make for some crazy plumbing.
- Alternatively, eat the cost of runtime casts and funnel it through `type_caster<>`. That's what we do for our fork ([`RobotLocomotion/pybind11`](https://github.com/RobotLocomotion/pybind11/blob/58a368ea8e89638e01a7385cad9ce4345d70003d/README_DRAKE.md)), and we haven't yet had to point fingers at performance there (though our use case may be simpler).

@YannickJadoul or @rgwk Any chance y'all have a good (and mebbe easy?) timing performance benchmark for this, to see what the risk is in these terms?

Reply:
@EricCousineau-TRI wrote:
> Any chance y'all have a good (and mebbe easy?) timing performance benchmark for this, to see what the risk is in these terms?

The TensorFlow core team has very sophisticated pybind11 benchmarks, but it's currently Google-internal only. The author already gave me permission to extract most of it for external view, including sources, but it may take me a few days (I want to show what I extract to the author for approval).

Reply:
Sweet!!! That'd be awesome! I'll see if I can gather some common "complex" patterns in Drake to see if I can get some patterns we have (but maintain feature-parity with upstream).

Have you thought any about where your benchmarks might live? I think the ones I'd generate would be simple enough (i.e. just `pybind11` bits) to be in a benchmarking subdirectory.

Reply:
I explicitly converted my previous code doing compatibility checks at cast time to this code, doing these tests at definition time only, because of efficiency concerns raised elsewhere.